### ISSUE TITLE
Undetected Fortran incompatibility?

### ISSUE THREAD

**Issue by eschnett** [JuliaPackaging/Yggdrasil#5246]
I am trying to build `MSTM` (see https://dev.azure.com/JuliaPackaging/Yggdrasil/_build/results?buildId=20987&view=logs&j=572c5e49-83d5-5271-390a-e6dc77f89c6b&t=25aa3cdb-e20d-599e-657c-8776cc8a525f). The architecture `i686-linux-gnu-libgfortran3-mpich` (and many other architectures) fails to build with the error message

```
Fatal Error: Cannot read module file ‘mpi.mod’ opened at (1), because it was created by a different version of GNU Fortran
```

`MPICH_jll` contains Fortran modules (`*.mod` files, e.g. `mpi.mod`). These are part of the Fortran 90 MPI interface. Such module files are usually compiler- and version-specific.

I assume this means that `MSTM` needs to be built with the same (or nearly the same) Fortran compiler as `MPICH`. Both packages do `expand_gfortran_versions`, but apparently that's not enough?

**Comment by eschnett** [comment]
It turns out that OpenMPI does not have this problem, only MPICH and MPItrampoline do. I checked for related differences in their build scripts, and the only thing I can spot is that OpenMPI says `preferred_gcc_version=v"5"`, whereas MPICH and MPItrampoline don't require any specific GCC version.

MSTM already has a work-around for this problem for Windows. There, the file `mpi.f90` is apparently shipped with the MPI implementation, and the MSTM build script simply recompiles it. I tried the same approach for MPItrampoline, and this works (i.e. builds). It's more difficult for MPICH because the file `mpi.f90` there cannot be downloaded: it is generated during the build process, but is not installed.

Would it be worthwhile to try setting `preferred_gcc_version=v"5"` for MPICH to see whether this helps? Unfortunately it's difficult for me to test this because BinaryBuilder doesn't seem to want to use packages that only exist locally on my file system, and at the same time the authentication mechanism that lets me deploy to Github isn't working any more (unless it was recently fixed).
